src/metabase/api/collection.clj

```diff
@@ -402,7 +406,15 @@
 
 (defmethod post-process-collection-children :dataset
   [_ rows]
-  (post-process-collection-children :card rows))
+  (let [dataset_queries (map :dataset_query rows)]
+    (->> rows
+         ;; normalize dataset queries just for based_on_upload hydration
+         (map #(update % :dataset_query (comp mbql.normalize/normalize json/parse-string)))
+         upload/model-hydrate-based-on-upload
+         (map (fn [dataset_query row]
+                (assoc row :dataset_query dataset_query))
+              dataset_queries)
```

Comment:
This "normalize then undo" pattern could be made clearer IMHO. I see this has already shipped, so I'll send another PR for it.